### PR TITLE
All fetching logic complete, including content signature checks

### PR DESCRIPTION
**Pull request by leogrim**
@stkem @eishay @andrewconner @yingjieMiao 
I still have to make `RoverDocumentFetcher` use proxies, but I'm going to work on the queues / scheduling first.

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/rover/store/ArticleStore.scala [hunk 1]
```diff
@@ -9,16 +11,45 @@ import com.keepit.rover.article.{ Article, ArticleKind }
...
+  // This wrapper enforces consistency and provides APIs with more specific typing.
+
+  def get[A <: Article](key: ArticleKey[A])(implicit classTag: ClassTag[A]): Future[Option[A]] = {
+    SafeFuture {
+      underlying.get(key).map {
+        case expectedArticle: A => expectedArticle
+        case unexpectedArticle => throw new InconsistentArticleTypeException(key, unexpectedArticle)
+      }
+    }
+  }
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/rover/store/ArticleStore.scala:46]: @andrewconner couldn't get it to work with kind :(

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/rover/store/ArticleStore.scala:46]: Aw

File: kifi-backend/modules/common/app/com/keepit/rover/model/BasicArticleInfo.scala [hunk 2]
```diff
@@ -19,15 +19,26 @@ case class ArticleVersion(major: VersionNumber[Article], minor: VersionNumber[Ar
...
+  def uriId: Id[NormalizedURI]
   def bestVersion: Option[ArticleVersion]
   def latestVersion: Option[ArticleVersion]
-  def articleKind = ArticleKind.byTypeCode(kind)
-  def getLatestKey: Option[ArticleKey] = bestVersion.map(toKey)
-  def getBestKey: Option[ArticleKey] = latestVersion.map(toKey)
-  private def toKey(version: ArticleVersion) = ArticleKey(uriId, articleKind, version)
+  def getLatestKey[A <: Article](implicit kind: ArticleKind[A]): Option[ArticleKey[A]] = bestVersion.map(toKey(_))
+  def getBestKey[A <: Article](implicit kind: ArticleKind[A]): Option[ArticleKey[A]] = latestVersion.map(toKey(_))
+  private def toKey[A <: Article](version: ArticleVersion)(implicit kind: ArticleKind[A]) = ArticleKey(uriId, kind, version)
```
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/rover/model/BasicArticleInfo.scala:41]: `bestVersion` and `latestVersion` swapped?

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/rover/model/BasicArticleInfo.scala:41]: Yes! Good Catch! I remember fixing it but I must have stashed that :)

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/rover/model/BasicArticleInfo.scala:41]: We should have a "everyone, check your stashes for uncommitted work" day. I'm sure I have a ton.

Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/rover/model/BasicArticleInfo.scala:41]: I saw you fixed it in a different file ;-)

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/rover/model/BasicArticleInfo.scala:41]: Ah :)
I have fixed this one in my next branch.

**Pull request merged**